projects/ad405x_iio/app/ad405x_iio.c

```diff
@@ -0,0 +1,1576 @@
+{
+	uint32_t timeout = BUF_READ_TIMEOUT;
+	uint32_t sample_index = 0;
+	int32_t data_read;
+	int32_t ret;
+	data_ready = false;
+	uint16_t local_tx_data = 0;
+	uint16_t spirxdma_ndtr;
+
+	nb_of_samples = iio_dev_data->buffer->size / BYTES_PER_SAMPLE;
+
+	nb_of_samples_g = nb_of_samples;
+	iio_dev_data_g = iio_dev_data;
+
+	if (!buf_size_updated) {
+		/* Update total buffer size according to bytes per scan for proper
+		 * alignment of multi-channel IIO buffer data */
+		iio_dev_data->buffer->buf->size = iio_dev_data->buffer->size;
+		buf_size_updated = true;
+	}
+
+#if (INTERFACE_MODE == SPI_INTERRUPT)
+	ret = ad405x_start_data_capture();
+	if (ret) {
+		return ret;
+	}
+
+	while (sample_index < nb_of_samples ) {
+		while (data_ready != true && timeout > 0) {
+			timeout--;
+		}
+
+		if (!timeout) {
+			return -EIO;
+		}
+
+		ret = ad405x_spi_data_read(p_ad405x_dev, &data_read);
+		if (ret) {
+			return ret;
+		}
+
+		ret = no_os_cb_write(iio_dev_data->buffer->buf, &data_read, BYTES_PER_SAMPLE);
+		if (ret) {
+			return ret;
+		}
+
+		sample_index++;
+		data_ready = false;
+	}
+
+	ret = ad405x_stop_data_capture();
+	if (ret) {
+		return ret;
+	}
+
+#else
+#if (APP_CAPTURE_MODE == WINDOWED_DATA_CAPTURE)
+	ret = no_os_cb_prepare_async_write(iio_dev_data->buffer->buf,
+					   nb_of_samples*(BYTES_PER_SAMPLE), &buff_start_addr, &data_read);
+	if (ret) {
+		return ret;
+	}
+
+	if (!dma_config_updated) {
+		/* Cap SPI RX DMA NDTR to MAX_DMA_NDTR.
+		 * If number of bytes is more than that register the M1 callback, enable DBM*/
+		spirxdma_ndtr = no_os_min(MAX_DMA_NDTR, nb_of_samples * (BYTES_PER_SAMPLE));
+
+		if (nb_of_samples * (BYTES_PER_SAMPLE) > spirxdma_ndtr) {
+			/* Set flag indicating more than 64k samples requested
+			 * Enable double buffering
+			 * Register M1 Complt Callback
+			 * Specify M1 starting address*/
+			more_than_64k_requested = true;
+			hspi1.hdmarx->Instance->CR |= (uint32_t)(DMA_SxCR_DBM);
+			HAL_DMA_RegisterCallback(&hdma_spi1_rx,
+						 HAL_DMA_XFER_M1CPLT_CB_ID,
+						 rxM1cmplt_callback);
+
+
+			hspi1.hdmarx->Instance->M1AR = (uint32_t)buff_start_addr + 0xffff;
+		} else {
+			more_than_64k_requested = false;
+			hspi1.hdmarx->Instance->CR &= ~(uint32_t)(DMA_SxCR_DBM);
+			HAL_DMA_UnRegisterCallback(&hdma_spi1_rx,
+						   HAL_DMA_XFER_M1CPLT_CB_ID);
+		}
+
+		struct no_os_spi_msg ad405x_spi_msg = {
+			.tx_buff = (uint32_t*)local_tx_data,
+			.rx_buff = (uint32_t*)buff_start_addr,
+			.bytes_number = spirxdma_ndtr
+		};
+
+		struct stm32_spi_desc* sdesc = p_ad405x_dev->spi_desc->extra;
+		ret = no_os_spi_transfer_dma_async(p_ad405x_dev->spi_desc,
+						   &ad405x_spi_msg,
+						   1,
+						   (void *)receivecomplete_callback,
+						   NULL);
+		if (ret) {
+			return ret;
+		}
+
+		/* Disable CS PWM and reset the counters */
+		no_os_pwm_disable(sdesc->pwm_desc); // CS PWM
+		htim2.Instance->CNT = 0;
+		htim1.Instance->CNT = 0;
+		TIM8->CNT = 0;
+
+		/* if more than 64k requested, enable DBM */
```

Comment:
Are we missing if here?

Reply:
The comment is not applicable anymore, so will remove it. We use DBM always to avoid having the first sample overwritten